firebase-firestore/src/main/java/com/google/firebase/firestore/UserDataConverter.java

```diff
@@ -183,40 +184,39 @@ public ParsedUpdateData parseUpdateData(List<Object> fieldsAndValues) {
   public FieldValue parseQueryValue(Object input) {
     ParseAccumulator accumulator = new ParseAccumulator(UserData.Source.Argument);
 
-    @Nullable FieldValue parsed = parseData(input, accumulator.rootContext());
+    @Nullable FieldValue parsed = convertAndParseData(input, accumulator.rootContext());
     hardAssert(parsed != null, "Parsed data should not be null.");
     hardAssert(
         accumulator.getFieldTransforms().isEmpty(),
         "Field transforms should have been disallowed.");
     return parsed;
   }
 
-  /**
-   * Converts a POJO into a Map, throwing appropriate errors if it wasn't actually a proper POJO.
-   */
-  public Map<String, Object> convertPOJO(Object pojo) {
-    checkNotNull(pojo, "Provided data must not be null.");
-    String reason =
+  /** Converts a POJO to native types and then parses it into model types. */
+  private FieldValue convertAndParseData(Object input, ParseContext context) {
+    UserData.Source source = context.getDataSource();
+    String badDocReason =
         "Invalid data. Data must be a Map<String, Object> or a suitable POJO object, but it was ";
 
     // Check Array before calling CustomClassMapper since it'll give you a confusing message
-    // to use List instead, which also won't work.
-    if (pojo.getClass().isArray()) {
-      throw new IllegalArgumentException(reason + "an array");
+    // to use List instead, which also won't work in a set().
+    if (source == UserData.Source.Set && input.getClass().isArray()) {
+      throw new IllegalArgumentException(badDocReason + "an array");
     }
 
-    Object converted = CustomClassMapper.convertToPlainJavaTypes(pojo);
-    if (!(converted instanceof Map)) {
-      throw new IllegalArgumentException(reason + "of type: " + Util.typeName(pojo));
+    // Convert POJOs
+    Object converted = CustomClassMapper.convertToPlainJavaTypes(input);
+
+    if (source == UserData.Source.Set && !(converted instanceof Map)) {
```

Comment:
How excited do you feel about moving this assert (and maybe the if-check in 203) to `parseMergeData` and `parseSetData`?

Reply:
I made a convertAndParseDocumentData() helper instead. Does that work?  This added 3 lines, but I found some test comments to fix up which ended up removing 7 lines, so we're okay.